src/ivc/step_circuit.rs

```diff
@@ -247,13 +258,47 @@ where
                     .try_into()
                     .unwrap();
 
-                Ok((next_step, new_U, assigned_input))
+                Ok((next_step_i, new_U, assigned_input))
             },
         )?;
 
-        let _output = self.synthesize_step(config.step_config, layouter, &input)?;
+        let z_output = self.synthesize_step(config.step_config, layouter, &assigned_input)?;
+
+        let mut ro_circuit = RO::new(
+            config.main_gate_config.clone(),
+            input.step_public_params.ro_constant.clone(),
+        );
+
+        let (x, y) = WrapValue::from_assigned_point(&assigned_input_witness.public_params_hash);
+        ro_circuit.absorb_point(x, y);
+
+        ro_circuit.absorb_base(WrapValue::Assigned(assigned_next_step_i));
```

Comment:
NBD: use a method to pack these absorbs together

Reply:
Too much ctx, more than 9 params to separate fn. I tried to make the ro_circuit job more concise, see how you like it now.

Reply:
LGTM